File: src/apps/daemon/internal/networking/dns/tunnel_probe_evaluator_test.go
package dns

import (
	"testing"
)

func TestComputeProbeMetrics(t *testing.T) {
	tcp := TCPProbeResult{
		Success: true,
		Attempts: []AttemptMetric{
			{Success: true, DurationMs: 50},
			{Success: true, DurationMs: 70},
			{Success: true, DurationMs: 90},
		},
		MedianRTTMs: 70,
		Consistency: 1.0,
	}
	udp := UDPProbeResult{
		Reachable: true,
		Attempts: []AttemptMetric{
			{Success: true, DurationMs: 60},
		},
	}
	dns := DNSProbeResult{
		UDPResponsive: true,
		TCPResponsive: true,
		Answers:       []string{"1.1.1.1"},
		Attempts: []AttemptMetric{
			{Success: true, DurationMs: 80},
		},
	}

	metrics := ComputeProbeMetrics(tcp, udp, dns)
	if metrics.RTTMs != 70 {
		t.Fatalf("expected RTTMs 70, got %d", metrics.RTTMs)
	}
	if metrics.JitterMs != 14 {
		t.Fatalf("expected JitterMs 14, got %d", metrics.JitterMs)
	}
	if metrics.PacketLossEstimate != 0.0 {
		t.Fatalf("expected PacketLossEstimate 0.0, got %f", metrics.PacketLossEstimate)
	}
	if metrics.StabilityPercent != 100.0 {
		t.Fatalf("expected StabilityPercent 100.0, got %f", metrics.StabilityPercent)
	}
}

func TestScoreProbeResult_TunnelReady(t *testing.T) {
	tcp := TCPProbeResult{
		Success:     true,
		Attempts:    []AttemptMetric{{Success: true, DurationMs: 40}},
		MedianRTTMs: 42,
		Consistency: 1.0,
	}
	tls := TLSProbeResult{
		Success:     true,
		HandshakeMs: 80,
		Version:     "TLS 1.3",
		Verified:    true,
	}
	http := HTTPProbeResult{
		Success: true,
		Probes: []HTTPProbeItem{
			{Method: "GET", URL: "https://example.com/", StatusCode: 200, DurationMs: 60},
		},
	}
	ws := WSProbeResult{
		Success:    true,
		StatusCode: 101,
		DurationMs: 70,
	}
	quic := QUICProbeResult{
		Success:     true,
		HandshakeMs: 65,
		ALPN:        "h3",
	}
	dns := DNSProbeResult{
		UDPResponsive: true,
		TCPResponsive: true,
		Answers:       []string{"1.1.1.1"},
	}
	metrics := ProbeMetrics{
		RTTMs:            42,
		JitterMs:         10,
		StabilityPercent: 98.0,
	}

	score := ScoreProbeResult(tcp, tls, http, ws, quic, dns, metrics)
	if score.Classification != "Tunnel Ready" {
		t.Fatalf("expected 'Tunnel Ready', got %s", score.Classification)
	}
	if score.Grade != "A+" {
		t.Fatalf("expected grade 'A+', got %s", score.Grade)
	}
	if score.FalsePositive {
		t.Fatalf("did not expect false positive")
	}
	if score.Numeric < 94 {
		t.Fatalf("expected numeric score >= 94, got %d", score.Numeric)
	}
}

func TestScoreProbeResult_FalsePositive_Spoofing(t *testing.T) {
	// TCP SYN-ACK received, but higher application layers blocked
	tcp := TCPProbeResult{
		Success:     true,
		Attempts:    []AttemptMetric{{Success: true, DurationMs: 20}},
		MedianRTTMs: 20,
		Consistency: 1.0,
	}
	tls := TLSProbeResult{Success: false}
	http := HTTPProbeResult{Success: false}
	ws := WSProbeResult{Success: false}
	quic := QUICProbeResult{Success: false}
	dns := DNSProbeResult{UDPResponsive: false, TCPResponsive: false}
	metrics := ProbeMetrics{RTTMs: 20, JitterMs: 2, StabilityPercent: 100.0}

	score := ScoreProbeResult(tcp, tls, http, ws, quic, dns, metrics)
	if !score.FalsePositive {
		t.Fatalf("expected false positive detection for spoofed TCP ACK")
	}
	if score.Classification != "False Positive" {
		t.Fatalf("expected classification 'False Positive', got %s", score.Classification)
	}
}

func TestScoreProbeResult_FalsePositive_InconsistentTCP(t *testing.T) {
	tcp := TCPProbeResult{
		Success:     true,
		Attempts:    []AttemptMetric{{Success: true, DurationMs: 20}},
		Consistency: 0.33,
	}
	score := ScoreProbeResult(tcp, TLSProbeResult{Success: true}, HTTPProbeResult{}, WSProbeResult{}, QUICProbeResult{}, DNSProbeResult{}, ProbeMetrics{})
	if !score.FalsePositive {
		t.Fatalf("expected false positive detection for inconsistent TCP")
	}
}

func TestParseCloudflareTrace(t *testing.T) {
	trace := "fl=123f45\nip=198.51.100.42\nloc=us\nts=1690000000\n"
	ip, loc := ParseCloudflareTrace(trace)
	if ip != "198.51.100.42" {
		t.Fatalf("expected ip '198.51.100.42', got %s", ip)
	}
	if loc != "US" {
		t.Fatalf("expected loc 'US', got %s", loc)
	}
}

File: src/packages/lumicore/tests/tunnel_probe_evaluator_test.rs
use lumicore::scanner::tunnel_probe_evaluator::{
    AttemptMetric, DnsProbeResult, HttpProbe, HttpProbeResult, ProbeMetrics, QuicProbeResult,
    ScoreResult, TcpProbeResult, TlsProbeResult, TunnelProbeEvaluator, TunnelScanReport,
    UdpProbeResult, WsProbeResult,
};

#[test]
fn test_compute_metrics_jitter_and_stability() {
    let tcp = TcpProbeResult {
        success: true,
        attempts: vec![
            AttemptMetric::ok(50),
            AttemptMetric::ok(70),
            AttemptMetric::ok(90),
        ],
        median_rtt_ms: 70,
        consistency: 1.0,
        error_category: None,
    };
    let udp = UdpProbeResult {
        reachable: true,
        attempts: vec![AttemptMetric::ok(60)],
        error_category: None,
    };
    let dns = DnsProbeResult {
        udp_responsive: true,
        tcp_responsive: true,
        answers: vec!["1.1.1.1".to_string()],
        attempts: vec![AttemptMetric::ok(80)],
        error_category: None,
    };

    let metrics = TunnelProbeEvaluator::compute_metrics(&tcp, &udp, &dns);
    assert_eq!(metrics.rtt_ms, 70);
    // 5 successful attempts: 50, 70, 90, 60, 80. Mean = 70.0.
    // Variances: (50-70)^2=400, (70-70)^2=0, (90-70)^2=400, (60-70)^2=100, (80-70)^2=100.
    // Sum = 1000 / 5 = 200. sqrt(200) ~= 14.14 -> 14.
    assert_eq!(metrics.jitter_ms, 14);
    assert_eq!(metrics.packet_loss_estimate, 0.0);
    assert_eq!(metrics.stability_percent, 100.0);
    assert_eq!(metrics.timeout_frequency, 0.0);
}

#[test]
fn test_score_probe_tunnel_ready() {
    let tcp = TcpProbeResult {
        success: true,
        attempts: vec![AttemptMetric::ok(40), AttemptMetric::ok(45)],
        median_rtt_ms: 42,
        consistency: 1.0,
        error_category: None,
    };
    let tls = TlsProbeResult {
        success: true,
        handshake_ms: 80,
        version: Some("TLS 1.3".to_string()),
        cipher_suite: Some("TLS_AES_128_GCM_SHA256".to_string()),
        alpn: Some("h2".to_string()),
        verified: true,
        error_category: None,
    };
    let http = HttpProbeResult {
        success: true,
        probes: vec![HttpProbe {
            method: "GET".to_string(),
            url: "https://example.com/".to_string(),
            status_code: 200,
            duration_ms: 60,
            redirected: false,
            error_category: None,
        }],
    };
    let ws = WsProbeResult {
        success: true,
        status_code: 101,
        duration_ms: 70,
        error_category: None,
    };
    let quic = QuicProbeResult {
        success: true,
        handshake_ms: 65,
        alpn: Some("h3".to_string()),
        error_category: None,
    };
    let dns = DnsProbeResult {
        udp_responsive: true,
        tcp_responsive: true,
        answers: vec!["1.1.1.1".to_string()],
        attempts: vec![AttemptMetric::ok(30)],
        error_category: None,
    };
    let metrics = ProbeMetrics {
        rtt_ms: 42,
        jitter_ms: 10,
        packet_loss_estimate: 0.0,
        stability_percent: 98.0,
        timeout_frequency: 0.0,
    };

    let score = TunnelProbeEvaluator::score_probe(
        &tcp, &tls, &http, &ws, &quic, &dns, &metrics,
    );

    assert_eq!(score.classification, "Tunnel Ready");
    assert_eq!(score.grade, "A+");
    assert!(!score.false_positive);
    assert!(score.numeric >= 94);
    assert!(score.confidence > 90.0);
}

#[test]
fn test_middlebox_spoofing_false_positive_detection() {
    // Middlebox returns TCP SYN-ACK, but drops/resets all actual application payload
    let tcp = TcpProbeResult {
        success: true,
        attempts: vec![AttemptMetric::ok(20), AttemptMetric::ok(22), AttemptMetric::ok(21)],
        median_rtt_ms: 21,
        consistency: 1.0,
        error_category: None,
    };
    let tls = TlsProbeResult { success: false, ..Default::default() };
    let http = HttpProbeResult { success: false, probes: vec![] };
    let ws = WsProbeResult { success: false, ..Default::default() };
    let quic = QuicProbeResult { success: false, ..Default::default() };
    let dns = DnsProbeResult { udp_responsive: false, tcp_responsive: false, ..Default::default() };

    let metrics = ProbeMetrics {
        rtt_ms: 21,
        jitter_ms: 1,
        packet_loss_estimate: 0.0,
        stability_percent: 100.0,
        timeout_frequency: 0.0,
    };

    let score = TunnelProbeEvaluator::score_probe(
        &tcp, &tls, &http, &ws, &quic, &dns, &metrics,
    );

    assert!(score.false_positive);
    assert_eq!(score.classification, "False Positive");
}

#[test]
fn test_low_consistency_false_positive() {
    // Inconsistent TCP (consistency < 0.67) indicates erratic RST/ACK injection
    let tcp = TcpProbeResult {
        success: true,
        attempts: vec![
            AttemptMetric::ok(30),
            AttemptMetric::fail(1000, "timeout"),
            AttemptMetric::fail(1000, "connection_refused"),
        ],
        median_rtt_ms: 30,
        consistency: 0.33,
        error_category: None,
    };
    let tls = TlsProbeResult { success: true, ..Default::default() };
    let http = HttpProbeResult::default();
    let ws = WsProbeResult::default();
    let quic = QuicProbeResult::default();
    let dns = DnsProbeResult::default();
    let metrics = ProbeMetrics::default();

    let score = TunnelProbeEvaluator::score_probe(
        &tcp, &tls, &http, &ws, &quic, &dns, &metrics,
    );

    assert!(score.false_positive);
    assert_eq!(score.classification, "False Positive");
}

#[test]
fn test_cloudflare_trace_parsing() {
    let trace_raw = "fl=123f45\nh=www.cloudflare.com\nip=198.51.100.42\nts=1690000000\nvisit_scheme=https\nuag=WhiteDNS-Desktop\ncolo=FRA\nsliver=none\nhttp=http/2\nloc=de\ntls=TLSv1.3\nsni=plaintext\nwarp=off\ngateway=off\nrbi=off\nkex=X25519\n";

    let (ip, loc) = TunnelProbeEvaluator::parse_cloudflare_trace(trace_raw);
    assert_eq!(ip, Some("198.51.100.42".to_string()));
    assert_eq!(loc, Some("DE".to_string()));
}

#[test]
fn test_report_evaluation() {
    let mut report = TunnelScanReport {
        endpoint: "1.1.1.1:443".to_string(),
        host: "1.1.1.1".to_string(),
        port: 443,
        tcp: TcpProbeResult {
            success: true,
            attempts: vec![AttemptMetric::ok(50)],
            median_rtt_ms: 50,
            consistency: 1.0,
            error_category: None,
        },
        tls: TlsProbeResult {
            success: true,
            handshake_ms: 80,
            ..Default::default()
        },
        http: HttpProbeResult { success: true, probes: vec![] },
        websocket: WsProbeResult { success: true, ..Default::default() },
        udp: UdpProbeResult::default(),
        quic: QuicProbeResult::default(),
        dns: DnsProbeResult::default(),
        metrics: ProbeMetrics::default(),
        score: None,
    };

    TunnelProbeEvaluator::evaluate_report(&mut report);
    assert!(report.score.is_some());
    let score = report.score.unwrap();
    assert!(!score.false_positive);
    assert!(score.numeric > 50);
}
